dali/pipeline/operator/arg_helper.h

```diff
@@ -15,94 +15,170 @@
-#include <dali/pipeline/data/views.h>
 #include <memory>
 #include <string>
+#include <utility>
+#include <vector>
+#include "dali/core/geom/mat.h"
+#include "dali/core/geom/vec.h"
+#include "dali/pipeline/data/tensor.h"
+#include "dali/pipeline/data/views.h"
+#include "dali/pipeline/operator/argument.h"
+#include "dali/pipeline/operator/op_spec.h"
 
 namespace dali {
 
-template <typename T>
-class ArgValue {
- public:
-  ArgValue() = default;
-  ArgValue(ArgValue &&) = default;
-  inline ArgValue(const ArgValue &other) { *this = other; }
-  explicit inline ArgValue(const std::string &name) : name_(name) {
+template <int N>
+vec<N> as_vec(TensorView<StorageCPU, const float, 1> view) {
+  if (view.num_elements() == 1) {
+    return vec<N>(view.data[0]);
   }
+  assert(N == view.num_elements());
+  return *reinterpret_cast<const vec<N>*>(view.data);
+}
 
-  void Update(const OpSpec &spec, ArgumentWorkspace &ws, bool use_default = true) {
-    if (spec.HasTensorArgument(name_)) {
-      arg_input_ = &ws.ArgumentInput(name_);
-      gpu_dirty_ = true;
-      is_set_ = true;
-    } else {
-      is_set_ = spec.HasArgument(name_);
-      arg_input_ = nullptr;
-      if (use_default || is_set_)
-        value_ = spec.GetArgument<T>(name_);
-    }
+template <int N>
+vec<N> as_vec(TensorView<StorageCPU, const float, DynamicDimensions> view) {
+  return as_vec<N>(view.to_static<1>());
+}
+
+template <int N, int M>
+mat<N, M> as_mat(TensorView<StorageCPU, const float, 2> view) {
+  assert(N * M == view.num_elements());
+  return *reinterpret_cast<const mat<N, M>*>(view.data);
+}
+
+template <int N, int M>
+mat<N, M> as_mat(TensorView<StorageCPU, const float, DynamicDimensions> view) {
+  return as_mat<N, M>(view.to_static<2>());
+}
+
+template <int ndim>
+struct ArgShapeFromSize {
+  TensorShape<ndim> operator()(int64_t size) const {
+    throw std::logic_error(make_string("Cannot infer an ", ndim, "D shape from a flat size."));
+    return {};
   }
+};
 
-  ArgValue &operator=(ArgValue &&) = default;
-  inline ArgValue &operator=(const ArgValue &other) {
-    name_ = other.name_;
-    is_set_ = other.is_set_;
-    gpu_dirty_ = true;
-    gpu_.reset();
-    value_  = other.value_;
-    arg_input_ = other.arg_input_;
-    return *this;
+template <>
+struct ArgShapeFromSize<0> {
+  TensorShape<0> operator()(int64_t size) const {
+    DALI_ENFORCE(size == 1, make_string("Expected a scalar argument, got ", size, " values"));
+    return {};
   }
+};
+
+template <>
+struct ArgShapeFromSize<1> {
+  TensorShape<1> operator()(int64_t size) const {
+    return { size };
+  }
+};
 
-  inline bool IsInput() const noexcept { return arg_input_ != nullptr; }
-  inline bool IsSet() const noexcept { return is_set_; }
 
-  template <int ndim = DynamicDimensions>
-  inline const TensorView<StorageCPU, T, ndim> Tensor(int sample_index) const {
-    ASSERT(IsInput());
-    return view<T, ndim>(arg_input_[sample_index]);
+template <typename T, int ndim = 0>
+class ArgValue {
```

Comment:
Please add some documentation, how we configure this thing, and that it can be used to obtain views of arguments with some uniform shape required for all samples (or not in the other overload).